File: app/src/main/java/com/dhsdevelopments/potato/messages/RegistrationIntentService.java
package com.dhsdevelopments.potato.messages;

import android.app.IntentService;
import android.content.Intent;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import com.dhsdevelopments.potato.Log;
import com.dhsdevelopments.potato.PotatoApplication;
import com.dhsdevelopments.potato.R;
import com.dhsdevelopments.potato.clientapi.gcm.GcmRegistrationRequest;
import com.dhsdevelopments.potato.clientapi.gcm.GcmRegistrationResult;
import com.google.android.gms.gcm.GoogleCloudMessaging;
import com.google.android.gms.iid.InstanceID;
import retrofit.Call;
import retrofit.Response;

import java.io.IOException;

public class RegistrationIntentService extends IntentService
{
    public static final String ACTION_REGISTER = "com.dhsdevelopments.potato.gcm.REGISTER";
    public final String PREFS_KEY_GCM_REGISTERED = "gcmRegisterOk";

    /**
     * Creates an IntentService.  Invoked by your subclass's constructor.
     */
    public RegistrationIntentService() {
        super( "RegistrationIntentService" );
    }

    @Override
    protected void onHandleIntent( Intent intent ) {
        if( intent != null && intent.getAction().equals( ACTION_REGISTER ) ) {
            handleRegister();
        }
    }

    private void handleRegister() {
        try {
            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences( this );

            InstanceID instanceId = InstanceID.getInstance( this );
            String token = instanceId.getToken( getString( R.string.gcm_sender_id ), GoogleCloudMessaging.INSTANCE_ID_SCOPE, null );

            Log.d( "Got token: " + token );

            PotatoApplication app = PotatoApplication.getInstance( this );
            Call<GcmRegistrationResult> call = app.getPotatoApi().registerGcm( app.getApiKey(), new GcmRegistrationRequest( token ) );
            Response<GcmRegistrationResult> result = call.execute();
            if( !result.isSuccess() ) {
                if( result.code() == 503 ) {
                    Log.w( "GCM is disabled on the server" );
                }
                else {
                    Log.e( "Error when updating GCM key: " + result.code() + ", " + result.message() );
                }
            }
            else if( "ok".equals( result.body().result ) ) {
                SharedPreferences.Editor prefsEditor = prefs.edit();
                prefsEditor.putBoolean( PREFS_KEY_GCM_REGISTERED, true );
                prefsEditor.apply();
            }
        }
        catch( IOException e ) {
            Log.e( "Error when requesting token", e );
        }
    }
}

File: app/src/main/java/com/dhsdevelopments/potato/service/RegistrationIntentService.kt
package com.dhsdevelopments.potato.service

import android.app.IntentService
import android.content.Context
import android.content.Intent
import android.preference.PreferenceManager
import com.dhsdevelopments.potato.PotatoApplication
import com.dhsdevelopments.potato.clientapi.gcm.GcmRegistrationRequest
import com.dhsdevelopments.potato.common.DbTools
import com.dhsdevelopments.potato.common.Log
import com.google.firebase.iid.FirebaseInstanceId

class RegistrationIntentService : IntentService("RegistrationIntentService") {

    companion object {
        const val ACTION_REGISTER = "com.dhsdevelopments.potato.gcm.REGISTER"

        private const val PREFS_KEY_GCM_REGISTERED = "gcmRegisterOk"

        fun tokenUpdated(context: Context) {
            val updatedToken = FirebaseInstanceId.getInstance().token
            if(updatedToken != null) {
                val app = PotatoApplication.getInstance(context)
                if(app.isAuthenticated()) {
                    sendTokenToServer(context, updatedToken)
                }
            }
        }

        fun sendTokenToServer(context: Context, token: String) {
            val app = PotatoApplication.getInstance(context)
            val prefs = PreferenceManager.getDefaultSharedPreferences(context)

            val call = app.findApiProvider().makePotatoApi().registerGcm(app.findApiKey(), GcmRegistrationRequest(token, "gcm"))
            val result = call.execute()
            if (!result.isSuccessful) {
                if (result.code() == 503) {
                    Log.w("GCM is disabled on the server")
                } else {
                    Log.e("Error when updating GCM key: " + result.code() + ", " + result.message())
                }
            } else if ("ok" == result.body()!!.result) {
                val prefsEditor = prefs.edit()
                prefsEditor.putBoolean(PREFS_KEY_GCM_REGISTERED, true)
                prefsEditor.apply()

                // If this was a new registration, we need to clear the channel notification configuration
                if (result.body()!!.detail == "token_registered") {
                    DbTools.clearUnreadNotificationSettings(context)
                }
            } else {
                Log.e("Unexpected reply from gcm registration: ${result.body()!!.result}")
            }
        }
    }

    override fun onHandleIntent(intent: Intent?) {
        if (intent != null && intent.action == ACTION_REGISTER) {
            handleRegister()
        }
    }

    private fun handleRegister() {
        tokenUpdated(this)
    }
}
